### PR TITLE
Base rate sub system unit tests

### PR DESCRIPTION
**Pull request by salman-codewar**
## Purpose changes

Issue Ref: #346 

**Base Rate sub-system**

- [x] Added UploadBaseRate unit test
- [x] Added GetBaseRates unit test
- [x] Added GetBaseRate unit test
- [x] Added Repo functions unit test
- [x] Added Util functions unit test
- [x] Added validation functions unit tests

**How To?**

To run tests, navigate to `base-rate` dir and run below commands

```go
make test
```

**Demo**
BaseRateTest

**Coverage**
coverage

### CODE REVIEW

File: systems/data-plan/base-rate/pkg/server/rate_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package server
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/server/rate_test.go:1]: Good start. But these only test canonical/happy paths.  We can try to add sub tests for corner cases like trying to get a BaseRase that does not exist, et c...

File: systems/data-plan/base-rate/pkg/server/rate_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	 },nil)
+	
+	 s := NewBaseRateServer(baseRateRepo)
+	rate, err := s.GetBaseRates(context.TODO(), mockFilters)
+	assert.NoError(t, err)
+	assert.Equal(t, "Tycho crater", rate.Rates[0].Country)
+	assert.Equal(t, "LTE", rate.Rates[0].Lte)
+	baseRateRepo.AssertExpectations(t)
+}
+
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/server/rate_test.go:88]: We also need to have unit tests for the following: 
/pkg/db/*_repo.go files
/utils.go file

**Review by delvatt — COMMENTED** [review]
Good. But, we also need to have unit tests for the following:
/pkg/db/*_repo.go files
/utils.go file

File: systems/data-plan/base-rate/pkg/server/rate_test.go (hunk 1 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/server/rate_test.go:1]: Addressed

File: systems/data-plan/base-rate/pkg/server/rate_test.go (hunk 2 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/server/rate_test.go:133]: Addressed. Also added test for `pkg/validations`

File: systems/data-plan/base-rate/pkg/utils/utils_test.go [hunk 3]
```diff
@@ -35,14 +35,18 @@ func TestRateService_IsFutureDate(t *testing.T) {
 }
 
 func TestRateService_FetchData(t *testing.T) {
-	//Success case
 	mockFileUrl := "https://raw.githubusercontent.com/ukama/ukama/main/systems/data-plan/docs/template/template.csv"
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/utils/utils_test.go:30]: As discussed last time, as soon as we have a real  url  like "https://raw.githubusercontent.com/ukama/ukama/main/systems/data-plan/docs/template/template.csv" and are making a real network/http call to that url, then this is no more a unit test, but becomes an integration test. 
For now, lets just keep this one here, and I 'll propose to you guys in a couple of days an alternative, and then we will finally move this test (which is a good integration test) into the integration tests pakage.

**Review by delvatt — APPROVED** [review]

**Review by delvatt — COMMENTED** [review]
Let's address these last comments and we 'll be good to go!

File: systems/data-plan/base-rate/pkg/server/rate_test.go (hunk 1 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/server/rate_test.go:1]: Yea, make sense. Addressed the feedback

File: systems/data-plan/base-rate/pkg/db/rate_repo_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		})
+
+		assert.NoError(t, err)
+
+		rates, err := r.GetBaseRates("Tycho crater", "", "", "inter_mno_data")
+		assert.NoError(t, err)
+		err = mock.ExpectationsWereMet()
+		assert.NoError(t, err)
+		assert.NotNil(t, rates)
+	})
```
[the remark this answers is not in the document]
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/db/rate_repo_test.go:156]: @delvatt Test for Nil case is already written in server test. Do we need to test it again at repo level?

File: .gitignore [hunk 5]
```diff
@@ -19,3 +19,4 @@
 node_modules
```
[the remark this answers is not in the document]
Comment (salman-codewar) [reply .gitignore:19]: `node_modules` already exist in gitignore on main branch. Idk why it's showing this change on my branch.

File: systems/data-plan/base-rate/pkg/db/rate_repo_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (delvatt) [reply systems/data-plan/base-rate/pkg/db/rate_repo_test.go:156]: It depends on the "WHY" for the nil case in server test. A nil server case does not necessary mean  a Not Found event in the underlying repo.  But that was just a suggestion. We don't have to go into too much details for this simple case.

**Review by delvatt — APPROVED** [review]
